pysages/methods/improved_string.py

```diff
@@ -0,0 +1,301 @@
+    Implementation of the improved string method.
+    The order (ignoring second order terms with covariance matrix) as described in
+    J. Chem. Phys. 131, 034109 (2009); https://doi.org/10.1063/1.3175798 (equation 13).
+    Higher order approximations can be implemented by the user using the provided
+    covariance matrix.
+
+    Arguments
+    ---------
+    context_generator: Callable
+        User defined function that sets up a simulation context with the backend.
+        Must return an instance of `hoomd.conext.SimulationContext` for HOOMD-blue and
+        `openmm.Context` for OpenMM.
+        The function gets `context_args` unpacked for additional user args.
+        For each replica along the path, the argument `replica_num` in [0, ..., N-1]
+        is set in the `context_generator` to load the appropriate initial condition.
+
+    timesteps: int
+        Number of timesteps the simulation is running.
+
+    stringsteps: int
+       Number of steps the string positions are iterated.
+       It is the user responsibility to ensure final convergence.
+
+    context_args: Optional[dict] = None
+        Arguments to pass down to `context_generator` to setup the simulation context.
+
+    kwargs:
+        Passed to the backend run function as additional user arguments.
+
+    post_run_action: Optional[Callable] = None
+        Callable function that enables actions after the run execution of PySAGES.
+        Actions are executed inside the generated context.
+        Example uses for this include writing a final configuration file.
+        This function gets `context_args` unpacked just like `context_generator`.
+
+    * Note:
+        This method does not accept a user defined callback.
+    """
+    timesteps = int(timesteps)
+    stringsteps = int(stringsteps)
+    context_args = {} if context_args is None else context_args
+    cv_shape = np.asarray(method.cvs).shape
+
+    for step in range(stringsteps):
+        context_args["stringstep"] = len(method.path_history)
+        # Reset histograms before new simulation
+        for i in range(len(method.umbrella_sampler.histograms)):
+            if i not in method.freeze_idx:
+                method.umbrella_sampler.histograms[i].reset()
+        umbrella_kwargs = copy.deepcopy(kwargs)
+        umbrella_kwargs["executor_shutdown"] = False
+        umbrella_result = pysages.run(
+            method.umbrella_sampler,
+            context_generator,
+            timesteps,
+            context_args,
+            post_run_action,
+            executor,
+            **umbrella_kwargs
+        )
+
+        new_xi = []
+        for i in range(len(umbrella_result.callbacks)):
+            sampled_xi = umbrella_result.callbacks[i].get_means()
+            old_xi = method.umbrella_sampler.submethods[i].center
+            direction = (sampled_xi - old_xi) / method.metric(sampled_xi, old_xi)
+            new_xi.append(np.asarray(old_xi) + np.asarray(method.alpha) * np.asarray(direction))
```

Comment:
Can we call `np.asarray` over alpha during `__init__` and just use `method.alpha` here?

Reply:
sure: https://github.com/SSAGESLabs/PySAGES/pull/182/commits/6f51494602de49d506a61be11edcc9eea7871136